### ISSUE TITLE
Missing support for "timestamp with time zone"

### ISSUE THREAD

**Issue by wearpants** [snakedotdev/grafana-duckdb-datasource#16]
see attached. this works fine in CLI
<img width="1217" alt="Screenshot 2024-10-26 at 10 41 54 PM" src="https://github.com/user-attachments/assets/285cee61-4170-4242-a329-71d185d9ac9a">

**Comment by wearpants** [comment]
though from `.schema`

`CREATE TABLE sensors_raw(ts TIMESTAMP, device VARCHAR, temperature FLOAT, pressure FLOAT);`

**Comment by wearpants** [comment]
possibly related https://github.com/marcboeker/go-duckdb?tab=readme-ov-file#notes-and-faqs

**Comment by wearpants** [comment]
Actually yes current timestamp returns a time zone type

```D describe select current_timestamp;
┌───────────────────┬──────────────────────────┬─────────┬─────────┬─────────┬─────────┐
│    column_name    │       column_type        │  null   │   key   │ default │  extra  │
│      varchar      │         varchar          │ varchar │ varchar │ varchar │ varchar │
├───────────────────┼──────────────────────────┼─────────┼─────────┼─────────┼─────────┤
│ current_timestamp │ TIMESTAMP WITH TIME ZONE │ YES     │         │         │         │
└───────────────────┴──────────────────────────┴─────────┴─────────┴─────────┴─────────┘
```

**Comment by colearendt** [comment]
Very interesting! And a workaround would be to convert that to a `TIMESTAMP` beforehand, like:

```
(now()::timestamp) - INTERVAL '2 hour'
```

Although it's important to be careful of dropping the tz in that way (have to modify time filter accordingly)

It is strange that it works in the CLI though. I may have to try through some other client libraries as a sanity check for whether the CLI is doing niceties to workaround or whether this is a standard issue with odbc / other client libraries.

**Comment by colearendt** [comment]
Using the Python CLI, it looks like there may be some magic done in parsing the string:
```
>>> con.sql("SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;")
┌────────────────────────────────────────────────────────────────────────────┐
│ (CURRENT_TIMESTAMP - to_hours(CAST(trunc(CAST('2' AS DOUBLE)) AS BIGINT))) │
│                          timestamp with time zone                          │
├────────────────────────────────────────────────────────────────────────────┤
│ 2024-11-26 01:50:58.602-05                                                 │
└────────────────────────────────────────────────────────────────────────────┘
```

**Comment by colearendt** [comment]
The R client library (less popular) does not handle this elegantly. So there is at least some precedence for concluding that:

> This is _technically_ invalid duckdb syntax for which the python client library and CLI have implemented custom parsing logic to rewrite the query

```
dbExecute(con, "SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;")
Error in `dbSendQuery()`:
! rapi_prepare: Failed to prepare query SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
Error: Binder Error: No function matches the given name and argument types '-(TIMESTAMP WITH TIME ZONE, INTERVAL)'. You might need to add explicit type casts.
	Candidate functions:
	-(TINYINT) -> TINYINT
	-(TINYINT, TINYINT) -> TINYINT
	-(SMALLINT) -> SMALLINT
	-(SMALLINT, SMALLINT) -> SMALLINT
	-(INTEGER) -> INTEGER
	-(INTEGER, INTEGER) -> INTEGER
	-(BIGINT) -> BIGINT
	-(BIGINT, BIGINT) -> BIGINT
	-(HUGEINT) -> HUGEINT
	-(HUGEINT, HUGEINT) -> HUGEINT
	-(FLOAT) -> FLOAT
	-(FLOAT, FLOAT) -> FLOAT
	-(DOUBLE) -> DOUBLE
	-(DOUBLE, DOUBLE) -> DOUBLE
	-(DECIMAL) -> DECIMAL
	-(DECIMAL, DECIMAL) -> DECIMAL
	-(UTINYINT) -> UTINYINT
	-(UTINYINT, UTINYINT) -> UTINYINT
	-(USMALLINT) -> USMALLINT
	-(USMALLINT, USMALLINT) -> USMALLINT
	-(UINTEGER) -> UINTEGER
	-(UINTEGER, UINTEGER) -> UINTEGER
	-(UBIGINT) -> UBIGINT
	-(UBIGINT, UBIGINT) -> UBIGINT
	-(UHUGEINT) -> UHUGEINT
	-(UHUGEINT, UHUGEINT) -> UHUGEINT
	-(DATE, DATE) -> BIGINT
	-(DATE, INTEGER) -> DATE
	-(TIMESTAMP, TIMESTAMP) -> INTERVAL
	-(INTERVAL, INTERVAL) -> INTERVAL
	-(DATE, INTERVAL) -> TIMESTAMP
	-(TIME, INTERVAL) -> TIME
	-(TIMESTAMP, INTERVAL) -> TIMESTAMP
	-(TIME WITH TIME ZONE, INTERVAL) -> TIME WITH TIME ZONE
	-(INTERVAL) -> INTERVAL

LINE 1: SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
                                 ^
Run `rlang::last_trace()` to see where the error occurred.
```

As discussed above, by converting to Timestamp, this works. 

It also seems like this `-` operation exists if the `ICU` extension is enabled?

https://duckdb.org/docs/sql/functions/timestamptz.html#icu-timestamp-with-time-zone-operators

https://duckdb.org/docs/extensions/icu.html

This seems do-able perhaps...

```
# for macos, for example
curl -OL http://extensions.duckdb.org/v1.0.0/osx_arm64/icu.duckdb_extension.gz
```

```
INSTALL 'icu.duckdb_extension.gz';
LOAD icu
```

There is definitely some weirdness around versions. For instance, it seems like the R library gets its version from the hash of the version (i.e. SHA), which conflicts with the version tag and throws an error when loading extensions 😑 

Should probably test directly with the Golang library / etc. Can see platform stuff here:

https://duckdb.org/docs/extensions/working_with_extensions.html

**Comment by wearpants** [comment]
I'd be highly surprised if the python client and CLI (officially maintained) are doing custom parsing, I think the more likely root cause is that the Golang & R bindings (third party) have the same underlying problemSent from my iPhoneOn Nov 26, 2024, at 4:12 AM, Cole Arendt ***@***.***> wrote:﻿
The R client library (less popular) does not handle this elegantly. So there is at least some precedence for concluding that:

This is technically invalid duckdb syntax for which the python client library and CLI have implemented custom parsing logic to rewrite the query

dbExecute(con, "SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;")
Error in `dbSendQuery()`:
! rapi_prepare: Failed to prepare query SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
Error: Binder Error: No function matches the given name and argument types '-(TIMESTAMP WITH TIME ZONE, INTERVAL)'. You might need to add explicit type casts.
	Candidate functions:
	-(TINYINT) -> TINYINT
	-(TINYINT, TINYINT) -> TINYINT
	-(SMALLINT) -> SMALLINT
	-(SMALLINT, SMALLINT) -> SMALLINT
	-(INTEGER) -> INTEGER
	-(INTEGER, INTEGER) -> INTEGER
	-(BIGINT) -> BIGINT
	-(BIGINT, BIGINT) -> BIGINT
	-(HUGEINT) -> HUGEINT
	-(HUGEINT, HUGEINT) -> HUGEINT
	-(FLOAT) -> FLOAT
	-(FLOAT, FLOAT) -> FLOAT
	-(DOUBLE) -> DOUBLE
	-(DOUBLE, DOUBLE) -> DOUBLE
	-(DECIMAL) -> DECIMAL
	-(DECIMAL, DECIMAL) -> DECIMAL
	-(UTINYINT) -> UTINYINT
	-(UTINYINT, UTINYINT) -> UTINYINT
	-(USMALLINT) -> USMALLINT
	-(USMALLINT, USMALLINT) -> USMALLINT
	-(UINTEGER) -> UINTEGER
	-(UINTEGER, UINTEGER) -> UINTEGER
	-(UBIGINT) -> UBIGINT
	-(UBIGINT, UBIGINT) -> UBIGINT
	-(UHUGEINT) -> UHUGEINT
	-(UHUGEINT, UHUGEINT) -> UHUGEINT
	-(DATE, DATE) -> BIGINT
	-(DATE, INTEGER) -> DATE
	-(TIMESTAMP, TIMESTAMP) -> INTERVAL
	-(INTERVAL, INTERVAL) -> INTERVAL
	-(DATE, INTERVAL) -> TIMESTAMP
	-(TIME, INTERVAL) -> TIME
	-(TIMESTAMP, INTERVAL) -> TIMESTAMP
	-(TIME WITH TIME ZONE, INTERVAL) -> TIME WITH TIME ZONE
	-(INTERVAL) -> INTERVAL

LINE 1: SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
                                 ^
Run `rlang::last_trace()` to see where the error occurred.

As discussed above, by converting to Timestamp, this works.
It also seems like this - operation exists if the ICU extension is enabled?
https://duckdb.org/docs/sql/functions/timestamptz.html#icu-timestamp-with-time-zone-operators
https://duckdb.org/docs/extensions/icu.html
This seems do-able perhaps...
# for macos, for example
curl -OL http://extensions.duckdb.org/v1.0.0/osx_arm64/icu.duckdb_extension.gz

INSTALL 'icu.duckdb_extension.gz';
LOAD icu

There is definitely some weirdness around versions. For instance, it seems like the R library gets its version from the hash of the version (i.e. SHA), which conflicts with the version tag and throws an error when loading extensions 😑
Should probably test directly with the Golang library / etc. Can see platform stuff here:
https://duckdb.org/docs/extensions/working_with_extensions.html

**Comment by wearpants** [comment]
Can you clarify/Simplify what the issue is here? Does the golang binding just not properly support time zones at all or…?Sent from my iPhoneOn Nov 26, 2024, at 6:59 AM, Pete Fein ***@***.***> wrote:﻿I'd be highly surprised if the python client and CLI (officially maintained) are doing custom parsing, I think the more likely root cause is that the Golang & R bindings (third party) have the same underlying problemSent from my iPhoneOn Nov 26, 2024, at 4:12 AM, Cole Arendt ***@***.***> wrote:﻿
The R client library (less popular) does not handle this elegantly. So there is at least some precedence for concluding that:

This is technically invalid duckdb syntax for which the python client library and CLI have implemented custom parsing logic to rewrite the query

dbExecute(con, "SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;")
Error in `dbSendQuery()`:
! rapi_prepare: Failed to prepare query SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
Error: Binder Error: No function matches the given name and argument types '-(TIMESTAMP WITH TIME ZONE, INTERVAL)'. You might need to add explicit type casts.
	Candidate functions:
	-(TINYINT) -> TINYINT
	-(TINYINT, TINYINT) -> TINYINT
	-(SMALLINT) -> SMALLINT
	-(SMALLINT, SMALLINT) -> SMALLINT
	-(INTEGER) -> INTEGER
	-(INTEGER, INTEGER) -> INTEGER
	-(BIGINT) -> BIGINT
	-(BIGINT, BIGINT) -> BIGINT
	-(HUGEINT) -> HUGEINT
	-(HUGEINT, HUGEINT) -> HUGEINT
	-(FLOAT) -> FLOAT
	-(FLOAT, FLOAT) -> FLOAT
	-(DOUBLE) -> DOUBLE
	-(DOUBLE, DOUBLE) -> DOUBLE
	-(DECIMAL) -> DECIMAL
	-(DECIMAL, DECIMAL) -> DECIMAL
	-(UTINYINT) -> UTINYINT
	-(UTINYINT, UTINYINT) -> UTINYINT
	-(USMALLINT) -> USMALLINT
	-(USMALLINT, USMALLINT) -> USMALLINT
	-(UINTEGER) -> UINTEGER
	-(UINTEGER, UINTEGER) -> UINTEGER
	-(UBIGINT) -> UBIGINT
	-(UBIGINT, UBIGINT) -> UBIGINT
	-(UHUGEINT) -> UHUGEINT
	-(UHUGEINT, UHUGEINT) -> UHUGEINT
	-(DATE, DATE) -> BIGINT
	-(DATE, INTEGER) -> DATE
	-(TIMESTAMP, TIMESTAMP) -> INTERVAL
	-(INTERVAL, INTERVAL) -> INTERVAL
	-(DATE, INTERVAL) -> TIMESTAMP
	-(TIME, INTERVAL) -> TIME
	-(TIMESTAMP, INTERVAL) -> TIMESTAMP
	-(TIME WITH TIME ZONE, INTERVAL) -> TIME WITH TIME ZONE
	-(INTERVAL) -> INTERVAL

LINE 1: SELECT CURRENT_TIMESTAMP - INTERVAL '2' HOUR;
                                 ^
Run `rlang::last_trace()` to see where the error occurred.

As discussed above, by converting to Timestamp, this works.
It also seems like this - operation exists if the ICU extension is enabled?
https://duckdb.org/docs/sql/functions/timestamptz.html#icu-timestamp-with-time-zone-operators
https://duckdb.org/docs/extensions/icu.html
This seems do-able perhaps...
# for macos, for example
curl -OL http://extensions.duckdb.org/v1.0.0/osx_arm64/icu.duckdb_extension.gz

INSTALL 'icu.duckdb_extension.gz';
LOAD icu

There is definitely some weirdness around versions. For instance, it seems like the R library gets its version from the hash of the version (i.e. SHA), which conflicts with the version tag and throws an error when loading extensions 😑
Should probably test directly with the Golang library / etc. Can see platform stuff here:
https://duckdb.org/docs/extensions/working_with_extensions.html

**Comment by wearpants** [comment]
So the problem is that the `icu` extension is not built into golang (or R) bindings by default: https://github.com/marcboeker/go-duckdb/issues/228

See also: https://duckdb.org/docs/extensions/core_extensions.html#default-extensions

CLI:
```
D SELECT extension_name, installed, description FROM duckdb_extensions() where extension_name = 'icu';
'extension_name','installed','description'
'icu',true,'Adds support for time zones and collations using the ICU library'
```

Python:
```
>>> import duckdb
>>> duckdb.connect().execute("SELECT extension_name, installed, description from duckdb_extensions() where extension_name = 'icu';").fetchall()
[('icu', True, 'Adds support for time zones and collations using the ICU library')]
```

And for some reason, trying to query that table from within grafana errors:
<img width="1604" alt="Screenshot 2024-11-26 at 9 56 57 AM" src="https://github.com/user-attachments/assets/9c2a0022-dfaa-4906-8640-a3bd55d24aed">

**Comment by wearpants** [comment]
See upstream https://github.com/marcboeker/go-duckdb/issues/323 and https://github.com/marcboeker/go-duckdb/pull/280/

@colearendt If this requires significant work (such as building our own golang duckdb bindings instead of using prebuilts) I'm fine to put this on hold until end user demands it, as there's a clear solution identified.

**Comment by colearendt** [comment]
Did some more digging into this, as well as how it is done in the python library / CLI. I have a list of links and references somewhere... just need to find them!!

**Comment by wearpants** [comment]
closed in favor of root cause #19
